File: examples/nbody.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>
#include <shray2/shray.h>

#define min(a, b) ((a) < (b) ? a : b)

size_t flopCount = 0;

typedef struct {
    double x;
    double y;
    double z;
} Point;

void init(Point *positions, size_t n)
{
    for (size_t i = ShrayStart(n); i < ShrayEnd(n); i++) {
        positions[i].x = i;
        positions[i].y = i;
        positions[i].z = i;
    }
}

/* 16 flops */
Point accelerate(Point pos1, Point pos2, double mass)
{
    /* ||pos2 - pos1||^3 */
    double n = pow(pow(pos2.x - pos1.x, 2) + pow(pos2.y - pos1.y, 2) +
                pow(pos2.z - pos1.z, 2), 1.5);

    Point a;

    if (n == 0) {
        a.x = 0;
        a.y = 0;
        a.z = 0;
    } else {
        a.x = (pos2.x - pos1.x) * mass / n;
        a.y = (pos2.y - pos1.y) * mass / n;
        a.z = (pos2.z - pos1.z) * mass / n;
    }

    return a;
}

/* Calculates the acceleration of all locally owned bodies with respect to bodies
 * [start, end[. Blocks the loops with factor block. */
void accelerateHelp(Point *accel, Point *positions, double *masses, size_t n,
        size_t start, size_t end, size_t block)
{
    size_t localStart = ShrayStart(n);
    size_t localEnd = ShrayEnd(n);

    for (size_t Ib = localStart; Ib < localEnd; Ib += block) {
        for (size_t Jb = start; Jb < end; Jb += block) {
            for (size_t i = Ib; i < min(Ib + block, localEnd); i++) {
                for (size_t j = Jb; j < min(Jb + block, end); j++) {
                    Point acceleration = accelerate(positions[i], positions[j], masses[j]);
                    accel[i].x += acceleration.x;
                    accel[i].y += acceleration.y;
                    accel[i].z += acceleration.z;
                }
            }
        }
    }
}

void accelerateAll(Point *accel, Point *positions, double *masses, size_t n)
{
    size_t localStart = ShrayStart(n);
    size_t localEnd = ShrayEnd(n);
    unsigned int p = ShraySize();
    unsigned int s = ShrayRank();

    size_t block = 200;

    for (size_t i = localStart; i < localEnd; i++) {
        accel[i].x = 0.0;
        accel[i].y = 0.0;
        accel[i].z = 0.0;
    }

    ShrayPrefetch(&positions[(s + 1) % p * n / p], n / p * sizeof(Point));
    ShrayPrefetch(&masses[(s + 1) % p * n / p], n / p * sizeof(double));

    accelerateHelp(accel, positions, masses, n, localStart, localEnd, block);

    /* Accelerate with respect to P(t) and prefetch the next block (we pretend we have a loop, so
     * P(0) comes after P(p - 1)). */
    for (unsigned int t = (s + 1) % p; t != s; t = (t + 1) % p) {
        size_t startT = t * n / p;
        size_t endT = (t + 1) * n / p;

        if ((t + 1) % p != s) {
            size_t startNext = (t + 1) % p * n / p;
            ShrayPrefetch(&positions[startNext], n / p * sizeof(Point));
            ShrayPrefetch(&masses[startNext], n / p * sizeof(double));
        }

        accelerateHelp(accel, positions, masses, n, startT, endT, block);

        ShrayDiscard(&positions[startT], n / p * sizeof(Point));
        ShrayDiscard(&masses[startT], n / p * sizeof(double));
    }
}

/* Advances the n-bodies in place. accel is a buffer, does not need to be initialized. */
void advance(Point *positions, Point *velocities, double *masses,
        Point *accel, double dt, size_t n)
{
    accelerateAll(accel, positions, masses, n);
    ShraySync(accel);

    for (size_t i = ShrayStart(n); i < ShrayEnd(n); i++) {
        velocities[i].x += accel[i].x * dt;
        velocities[i].y += accel[i].y * dt;
        velocities[i].z += accel[i].z * dt;
        positions[i].x += velocities[i].x * dt;
        positions[i].y += velocities[i].y * dt;
        positions[i].z += velocities[i].z * dt;
    }
    ShraySync(positions);
    ShraySync(velocities);
}

int main(int argc, char **argv)
{
    ShrayInit(&argc, &argv);

    if (argc != 3) {
        printf("Usage: n, iterations\n");
        ShrayFinalize(1);
    }

    size_t n = atoll(argv[1]);
    if (n % ShraySize() != 0) {
        printf("Please make sure the number of nodes divides n.\n");
        ShrayFinalize(1);
    }
    int iterations = atoi(argv[2]);

    Point *positions = (Point *)ShrayMalloc(n, n * sizeof(Point));
    Point *velocities = (Point *)ShrayMalloc(n, n * sizeof(Point));
    double *masses = (double *)ShrayMalloc(n, n * sizeof(double));
    Point *accel = (Point *)ShrayMalloc(n, n * sizeof(Point));

    init(positions, n);
    ShraySync(positions);

    double duration;

    TIME(duration, for (int i = 0; i < iterations; i++) {
        advance(positions, velocities, masses, accel, 0.1, n);
    });

    if (ShrayOutput) {
        printf("%lf\n", 3.0 * 16.0 * n * n * iterations / 1000000000 / duration);
    }

    ShrayFree(positions);
    ShrayFree(velocities);
    ShrayFree(masses);
    ShrayFree(accel);

    ShrayReport();

    ShrayFinalize(0);

    return EXIT_SUCCESS;
}

File: examples/fortran/nbody.f90
! An n-body simulation. Still needs to be blocked

module accelerateModule
contains
function accelerate(pos1, pos2, mass) result(a)
    implicit none

    real(KIND=8), intent(in) :: pos1(3), pos2(3)
    real(KIND=8), intent(in) :: mass
    real(KIND=8) :: norm
    real(KIND=8) :: a(3)

    norm = ((pos2(1) - pos1(1)) ** 2 + (pos2(2) - pos1(2)) ** 2 + &
            (pos2(3) - pos1(3)) ** 2) ** 1.5

    a = (pos2 - pos1) * mass / norm
end function accelerate

subroutine accelerateAll(accel, positions, masses, n)
    implicit none

    integer, intent(in) :: n
    real(KIND=8) :: positions(n / num_images(), 3)[*]
    real(KIND=8) :: accel(n / num_images(), 3)[*]
    real(KIND=8) :: masses(n / num_images())[*]

    integer :: blockFactor, i, j, s
    integer :: I1, J1

    blockFactor = 100

    accel(:,:) = 0

    ! Unblocked
!    do s = 1, num_images()
!        do i = 1, n / num_images()
!            do j = 1, n / num_images()
!                accel(i,:) = accel(i, :) + &
!                                 accelerate(positions(i,:), positions(j,:)[s], masses(j)[s])
!            end do
!        end do
!    end do

    ! blocked
    do s = 1, num_images()
        do I1 = 1, n / num_images(), blockFactor
            do J1 = 1, n / num_images(), blockFactor
        do i = I1, MIN(I1 + blockFactor, n / num_images())
            ! Calculate acceleration with respect to processor s
            do j = J1, MIN(J1 + blockFactor, n / num_images())
                    accel(i,:) = accel(i,:) + &
                                 accelerate(positions(i,:), positions(j,:)[s], masses(j)[s])
            end do
        end do
            end do
        end do
    end do
end subroutine accelerateAll
end module accelerateModule


module nbodyModule
contains
subroutine advanceIt(positions, velocities, masses, accel, dt, n)
    use accelerateModule
    implicit none

    integer, intent(in) :: n
    real(KIND=8), intent(in) :: dt
    real(KIND=8) :: masses(n / num_images())[*]
    real(KIND=8) :: positions(n / num_images(), 3)[*], velocities(n / num_images(), 3)[*]
    real(KIND=8) :: accel(n / num_images(), 3)[*]

    call accelerateAll(accel, positions, masses, n)
    velocities = velocities + accel * dt
    positions = velocities * dt
    sync all

end subroutine advanceIt
end module nbodyModule

program main
    use nbodyModule
    implicit none

    integer :: n, iterations, p, t
    real(KIND=8), dimension(:,:), codimension[:], allocatable :: positions, velocities, accel
    real(KIND=8), dimension(:), codimension[:], allocatable :: masses
    real(KIND=8) :: dt
    integer :: cpu_count, cpu_count2, count_rate, count_max
    character(len=12), dimension(:), allocatable :: args

    allocate(args(2))
    call get_command_argument(1, args(1))
    read (unit=args(1), fmt=*) n

    call get_command_argument(2, args(2))
    read (unit=args(2), fmt=*) iterations

    p = num_images()

    if (modulo(n, p) /= 0) then
        write (*, *) 'Please make sure n divides p'
    end if

    allocate(positions(n / p, 3)[*])
    allocate(velocities(n / p, 3)[*])
    allocate(accel(n / p, 3)[*])
    allocate(masses(n / p)[*])

    positions = 1
    masses = 1
    velocities = 0

    dt = 0.1

    sync all
    call system_clock(cpu_count, count_rate, count_max)

    do t = 1, iterations
        call advanceIt(positions, velocities, masses, accel, dt, n)
    end do

    sync all

    call system_clock(cpu_count2, count_rate, count_max)

    if (this_image() .eq. 1) then
        write (*, *) 3.0 * 16.0 * n * n * iterations / 1000000000.0 / &
            (real(cpu_count2 - cpu_count) / count_rate)
    end if
end program main
